### DISCUSSION TITLE
Discussion for Catalog Access Groups Feature

### DISCUSSION THREAD

**Discussion by troycarpenter** [ampache/ampache#3244] [Show and tell]
Moving the discussion of feature request #2993 here as to not clutter up the feature request thread with implementation discussions.

First off, I have database changes to support this feature.  It will require two new tables and a new user attribute.

The first table is simply keeps track of the access group and the group name.  It comes with an initial group called DEFAULT.
The second table is a matrix of the access group number with a catalog number, and either true or false to indicate if that group has access to the catalog.  Upon upgrading to this feature, all catalogs will be enabled in the DEFAULT access group.  Admins can then modify which catalogs should be enabled or disabled in the DEFAULT group after the database upgrade.

Finally, an access group will be added to the user profile table and all users will initially be assigned to the DEFAULT group.

That should make it less obtrusive when the feature is introduced because currently everyone has access to all catalogs anyway (unless one is restricted).  This feature should make the catalog restriction feature obsolete since it can be accomplished by the new profiles.

Both the album and song tables include a field indicating which catalog to which they belong.  I expect that the SQL statement that generates a list albums or songs (or combination) can utilize the allowed catalogs stored in the new tables to filter results to only the allowed catalogs.

There will need to be some other background tricks, like if after the filters are applied, an artist doesn't have any albums/songs to display then also don't show the artist.

**Comment by troycarpenter** [comment]
The above logic will also need to be tied into the various APIs that also generate albums/songs for remote clients.

**Comment by troycarpenter** [comment]
I believe I have finally worked out all the database relations and foreign keys and other settings to make database management easier.

I have created the following two tables and one TRIGGER event:
catalog_access_group which is a simple two column table for group ID and group name.  There is a DEFAULT group with ID 1 which will be assigned to any new users (especially auto-created ones like from LDAP).  Deleting a group will cause the TRIGGER described below to run.

catalog_access is a matrix table where a row represents the access group ID, the catalog ID and if enabled or not for this access group.  the access group ID is tied to the first table, and the catalog ID is tied to the catalog table.  If any entry is deleted from one of those parent tables, then the corresponding rows are deleted here (i.e., if a catalog is deleted, then all rows that contained that catalogID will be deleted.  If an access group is deleted, then all rows that contain that access group will be deleted.  I also have a database upgrade trigger so that will populate the DEFAULT group with all catalogs which will closely mirror the access people have before this feature is activated.

The user table has been modified to have a catalog_access_group field for each user defining which profile will be applied to the user.  By default, users get the DEFAULT access group.  If the group to which they are assigned is deleted, they will revert to the DEFAULT group via the table trigger.

The TRIGGER event will trigger when a row is deleted from the catalog_access_group table and will reset the group to DEFAULT for any users who were assigned to the deleted access group.  I had to use a TRIGGER because the Innodb used on the user table does not allow for a foreign key to set a default value on deletion.

So far I have tested creating the new database from scratch (clean install) and the database upgrade routines for existing installations.

Now I need to work on the GUI for users to define their groups, and then the backend to apply the group as a filter to the data.

**Comment by troycarpenter** [comment]
Query to get allowed catalog IDs by user whose ID is 2:
SELECT catalog_id from catalog_access INNER JOIN user ON user.catalog_access_group = catalog_access.access_group_id WHERE user.id='2' AND catalog_access .enabled=1 ORDER BY catalog_id;

This query returned all albums across all catalogs to which the user has access:
SELECT id,name,album_artist,catalog  FROM album WHERE catalog IN (SELECT catalog_id FROM catalog_access INNER JOIN user ON user.catalog_access_group = catalog_access.access_group_id WHERE user.id='3' AND catalog_access .enabled=1);

Likewise this pulled all the songs to which a user should have access:
SELECT id,title,artist,catalog  FROM song WHERE catalog IN (SELECT catalog_id FROM catalog_access INNER JOIN user ON user.catalog_access_group = catalog_access.access_group_id WHERE user.id='2' AND catalog_access .enabled=1);

NOW I need to figure out how and where to incorporate these modifications into the actual queries.

**Comment by troycarpenter** [comment]
So while looking through the back end, I noticed there is a table called user_catalog which seems to have cross reference between a user ID and a catalog ID with many to many relationship.  Each row also has a "level" entry, but I don't know what that is.  I have populated it but it doesn't seem to do anything and the private catalog function doesn't use it either (that seems to be the "filter_user" field in the catalog table).  It seems like the first go at this feature, but to only use the one table isn't as flexible as the two table reference I'm using.

**Reply by lachlan-00** [reply]
I actually don't even know what that catalog table is used for. I think that's left over old ampache stuff.

Seems like that should be dropped

**Reply by troycarpenter** [reply]
Yeah, I see that it's close to what I've added for this feature, but I've not found anywhere where it is used.

**Reply by lachlan-00** [reply]
it's actually a very old feature from at least as early as 2013. I think it's been long removed so i'm going to drop that table
![image](https://user-images.githubusercontent.com/1305249/167067528-99cdc0ef-1fe8-4896-8067-677977178a6f.png)

**Reply by troycarpenter** [reply]
I can add it to my database/upgrade code if you want.

**Reply by troycarpenter** [reply]
Done.

**Comment by troycarpenter** [comment]
After a little break and some more digging, I found the place in src/Repository/Model/Catalog.php where a lot of the filtering can happen.  I also finally got the catalog browse drop down to only show the catalogs the user has access to.

I should also note that this feature will supersede the "restrict catalog to user" setting/feature since that functionality is now possible with this feature and more.

The fun part is going to be figuring how to filter out artists based on if the status of songs or albums in the appropriate catalogs.

**Comment by troycarpenter** [comment]
Ok, that was easier than I expected.
Here's what works right now.  Given a set of Catalog IDs, content not in that set is not shown.  Going down the menu on the left side:

Browse: all content type (album, songs, artists, etc)
Dashboards: Trending/Newest/Popular for albums/artists
Playlists: I have not verified playlist functionality yet.  Clearly a playlist with content from allowed catalogs will work, but what about a playlist that has songs from a restricted catalog.
Search: Only random search works properly
Information: All info seems to be filtered appropriately.

Also, the catalog restrictions are not controllable via the GUI yet.  Right now my testing is done with data already in the database tables.

TODO:
Filter the search results for Songs/Albums/Artists
Filter global search (the one at the top)
Create some playlists and be sure songs are filtered appropriately
Create GUIs to control the feature.

**Comment by troycarpenter** [comment]
BTW, it turns out most of the heavy lifting is done in get_user_filter() in src/Repository/Model/Catalog.php

And, as noted in the feature request thread, I need to get the same filtering in the Subsonic API.

**Reply by lachlan-00** [reply]
yeah, i think the hardest part is just finding what's missing the filter sql

**Reply by troycarpenter** [reply]
I'm having the hardest time figuring out the search rule building system (I really like the print_r() function...).  I just can't figure out where all that stuff gets turned into SQL.  If I can find that, then I can add the extra WHERE condition for the user.

**Reply by lachlan-00** [reply]
Search is probably the most confusing part of the entire codebase. (Mostly because of how much i've expanded it over the years.)

I need to split it up a but more like it was done in the edge branch but i'd diverged so much from the codebase by the time i got that far

The WHERE is set in the %type%_to_sql functions.
![image](https://user-images.githubusercontent.com/1305249/167067933-25e57cbe-4887-4938-9884-6336fa736868.png)

After the WHERE is set there are joins set
![image](https://user-images.githubusercontent.com/1305249/167067988-33f73f1f-fcf2-4fba-b9dd-162542691b45.png)

so it joins the object_type based on the catalog_filter and your user_id

So a user starts a search
sets the type
sets the rules and generates the sql
and then gets the results from that search

**Reply by troycarpenter** [reply]
Not to ask for things handed to me on a sliver platter, but what file is that code from?

**Reply by lachlan-00** [reply]
lol, that's what i'm here for. My first ampache pull request was a search enhancement so i spend a lot of time in there.

The search class handles it all. `src/Repository/Model/Search.php`

**Reply by troycarpenter** [reply]
Thanks.  It looks like I've already been in that file for other changes.  The cleanup of this commit is going to be fun given all the extra debug statements I've added to help see flows and data.

**Reply by lachlan-00** [reply]
comment them out as you get it stable and then try to keep only really important ones or future debug ones.

I keep the search debug commented because it's such a big class that it's always hard to work out issues

**Reply by troycarpenter** [reply]
I have to thank you for the "catalog_map" feature.  This is basically an expanded version of that feature and is making it easier because I just need to replace your original catalog_map clauses with the new expanded one.  I've done most of the searches now and it seems to be filtering as I would expect.

**Comment by troycarpenter** [comment]
With pointers from @lachlan-00, I have now completed the changes for the searching.  As of now the top search bar and the search section on the left will not return results in a catalog for which the user is not authorized.

I have also tried where an artist is in two separate catalogs.  When both catalogs are allowed, all albums appear under the artist.  When only one of the two are allowed, only the albums in the permitted catalog show up.

TODO:
~~Filter the search results for Songs/Albums/Artists~~
~~Filter global search (the one at the top)~~
Create some playlists and be sure songs are filtered appropriately
Create GUIs to control the feature.
Make sure all this works with the Subsonic API

**Reply by troycarpenter** [reply]
Follow-up...playlists did NOT have their entries removed IF those entries were in a prohibited catalog.  Will look at that tomorrow.

**Reply by troycarpenter** [reply]
Playlists are going to be a pain.  Playlists apparently can have all media types.  However, the media type is mentioned in the playlist_data table, and I've yet to figure out if the field of one table (playlist_data.object_type) can be used as a table name in another query.  I don't think it can.

Which means that I need to process each object_type separately so I can name the table for that object (song, album, video, tv, and whatever else it can be.)  

Right now I've got everything working for playlists of songs, but playless processing uses "items" .  I don't really use Ampache for anything other than songs, so this will be fun to fully test.

Also, I'm not adjusting the playlist track number...so for someone who has a playlist filtered, they will know because there will be gaps in the playlist numbering.  The counts are correct...

Another edge-case scenario will be what happens when a user creates a playlist from media in an approved catalog, then that access is taken away.  Right now I think the user would not be able to delete items that are no longer accessible.  That can be an improvement for later.

**Reply by troycarpenter** [reply]
@lachlan-00 : What types of object_types can be in a playlist?  I've only seen songs so far.  Even on an album page, when I add it to a playlist it adds all the songs, not a single entry for the album.

Is there something similar for video media?

**Reply by troycarpenter** [reply]
Trying to answer my own question...it looks like only song and video are object types for playlists...at lest from the garbage collection routine in src/Repository/Model/Playlist.php

**Reply by troycarpenter** [reply]
I see you can also have podcast_episode.

**Reply by troycarpenter** [reply]
So I've written SQL code for songs and podcast_episodes in playlists.  I left a default clause that was the old SQL code and debug trace in the default case to report when there is an unhandled object_type in a playlist.

Songs in a playlist that are in a catalog to which the user does not have access to do not show up in playlists.  The user will have some indication when this happens because the playlist track numbers will have holes (like it might be 1,2,5,7,11) instead of sequential.

**Reply by troycarpenter** [reply]
One last bug here, I haven't been able to figure out a lone SQL error that pops out when listing playlists on the dashboard.  The playlist still shows, but I have yet to track down the problem.

**Reply by lachlan-00** [reply]
Multiple objects can be in playlists but the ui/code only has options to put in songs.

So while things can be put in there it's really only songs that get added in practice.

It will be extended though.

**Reply by lachlan-00** [reply]
> One last bug here, I haven't been able to figure out a lone SQL error that pops out when listing playlists on the dashboard.  The playlist still shows, but I have yet to track down the problem.

Post the sql and it might be easy enough to id

**Reply by troycarpenter** [reply]
The SQL errors are in the next post a bit further down.  Both are happening in different places, but the error is the same about a column not found.

**Reply by troycarpenter** [reply]
FYI, I fixed the second one in the list, but the first one, which happens when showing playlists in the dashboard, happens on my unmodified 5.3.2 system as well.

**Comment by troycarpenter** [comment]
So as of today, I believe I have all the database and backend handling for the catalog access filter feature.  If that part wasn't hard enough, the next step is that I need to develop the GUIs for this feature.  This will involve multiple screens:

1. The user screen needs to be modified to allow a drop down to select the catalog access level for the user.  This is NOT the same as the User Access Level.  The new dropdown should appear on the add and edit user screens.
2. A new screen which allows for the creation of the catalog access level, and a matrix of the access levels (across the top), and catalogs down the side, populated by checkboxes to indicate which catalogs can be access by each access level.
3. Probably could use a link or a multi-select drop down from the add/edit catalog screen to also make the assignment while creating or editing a catalog directly.

It probably makes more sense to verify and do the Subsonic API since that won't involve GUI screens.

TODO:
~~Filter the search results for Songs/Albums/Artists~~
~~Filter global search (the one at the top)~~
~~Create some playlists and be sure songs are filtered appropriately~~
Create GUIs to control the feature.
Make sure all this works with the Subsonic API

**Reply by troycarpenter** [reply]
Need to investigate still...

SQL errors on playslists on the dashboard:
```
(Ampache\Module\System\Dba) -> Error_query SQL: SELECT `id` FROM `playlist` WHERE `last_update` >= '1651287448'  AND `playlist`.`id` IN (SELECT `playlist` FROM `playlist_data` LEFT JOIN `song` ON `playlist_data`.`object_id` = `song`.`id` AND `playlist_data`.`object_type` = 'song' LEFT JOIN `catalog_map` ON `catalog_map`.`object_type` = 'song'  AND `catalog_map`.`object_id` = `song`.`id` LEFT JOIN `catalog` ON `catalog_map`.`catalog_id` = `catalog`.`id` WHERE `catalog_map`.`object_type` = 'song' AND `catalog`.`id` IN (SELECT catalog_id FROM catalog_access INNER JOIN user ON user.catalog_access_group = catalog_access.access_group_id WHERE user.id=3 AND catalog_access.enabled=1)  GROUP BY `playlist_data`.`playlist`) UNION SELECT `object_id` FROM `object_count` WHERE `object_type` = 'playlist' AND `date` >= '1651287448'  AND `playlist`.`id` IN (SELECT `playlist` FROM `playlist_data` LEFT JOIN `song` ON `playlist_data`.`object_id` = `song`.`id` AND `playlist_data`.`object_type` = 'song' LEFT JOIN `catalog_map` ON `catalog_map`.`object_type` = 'song'  AND `catalog_map`.`object_id` = `song`.`id` LEFT JOIN `catalog` ON `catalog_map`.`catalog_id` = `catalog`.`id` WHERE `catalog_map`.`object_type` = 'song' AND `catalog`.`id` IN (SELECT catalog_id FROM catalog_access INNER JOIN user ON user.catalog_access_group = catalog_access.access_group_id WHERE user.id=3 AND catalog_access.enabled=1)  GROUP BY `playlist_data`.`playlist`) LIMIT 10 []
2022-05-07T02:57:28+00:00 [tu2] (Ampache\Module\System\Dba) -> Error_query MSG: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'playlist.id' in 'IN/ALL/ANY subquery'
```

Error when trying random Album search with default rule set, 5 item count:
```
(Ampache\Module\System\Dba) -> Error_query SQL: SELECT `song`.`id` FROM `song` WHERE `song`.`album` = ?  AND `album`.`id` IN (SELECT `object_id` FROM `catalog_map` LEFT JOIN `catalog` ON `catalog_map`.`catalog_id` = `catalog`.`id` WHERE `catalog_map`.`object_type` = 'album' AND `catalog`.`id` IN (SELECT catalog_id FROM catalog_access INNER JOIN user ON user.catalog_access_group = catalog_access.access_group_id WHERE user.id=3 AND catalog_access.enabled=1)  GROUP BY `catalog_map`.`object_id`)  ORDER BY `song`.`track`, `song`.`title` [172]
2022-05-07T03:01:14+00:00 [tu2] (Ampache\Module\System\Dba) -> Error_query MSG: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'album.id' in 'IN/ALL/ANY subquery'
```

**Reply by troycarpenter** [reply]
Found the problem with the second one.  Using an "album" filter for a song query.

**Reply by troycarpenter** [reply]
The first one is happening in "get_user_filter" for a playlist.  However, I put the original code back and it's still happening.  I went back to my unmodified Docker develop server and it's happening there too, so it look like it's an existing problem.  I will see if I can fix it.

**Reply by lachlan-00** [reply]
playlist isn't in the select here
```
SELECT `object_id`
FROM `object_count` 
WHERE `object_type` = 'playlist' AND 
      `date` >= '1651287448'  AND 
      `playlist`.`id`
```

**Reply by lachlan-00** [reply]
needs an object_count playlist
![image](https://user-images.githubusercontent.com/1305249/167319557-5e0c5dbe-0493-443f-91b1-9db104aa27f8.png)

**Reply by lachlan-00** [reply]
set with 6ec7acccd

**Reply by troycarpenter** [reply]
Confirmed fixed after pulling it into my branch.

**Comment by troycarpenter** [comment]
New bug found while testing...putting it here since I'm trying to get this code done:
Random Artist search not limited nor very random after the first three or 4 entries.

**Comment by troycarpenter** [comment]
I have now pushed the current backend handling code to my repository.  Of the two SQL errors mentioned, one is pre-existing and the second was in my new code.

Now on to the fun part of getting to the GUI...after Mother's day.

**Reply by troycarpenter** [reply]
I'm not going to look at this now (because I see it happens on an unmodified image) but it looks like playlists are available to everyone even if the list is private.  Probably need to look into that.

**Comment by troycarpenter** [comment]
After looking through the code, it appears that the APIs use the lower lever calls for most everything (good design work!)...which means that the Subsonic API just works with this feature.

To verify, I connected Dsub to the server using a "restricted" account, and on Dsub I only saw the catalogs/albumn/songs/etc that the user was allowed.

So now all that's left is the front-end web-GUI screens to provision this feature!

TODO:
~~Filter the search results for Songs/Albums/Artists~~
~~Filter global search (the one at the top)~~
~~Create some playlists and be sure songs are filtered appropriately~~
Create GUIs to control the feature.
~~Make sure all this works with the Subsonic API~~

**Comment by lachlan-00** [comment]
what do you want to create in the UI to manage?

**Reply by troycarpenter** [reply]
See next post.  I was typing it up when you asked.

**Comment by troycarpenter** [comment]
Here's what I'm thinking.  I've figured out how to put an entry in the menu for Catalog Access Groups.  I don't know if that's the best description, I'm also open to Catalog Filter Groups.  What I haven't gotten to yet is all the parameters passed down to make the various screens work.  It looks like code pulls in different templates depending on the command.  For expediency, if we can at least get control of the data from the GUI, we can improve the way it looks later.

For this feature, there are two main pieces of data.  There is an access group name and the access group matrix.  I'm thinking they can all be managed on the same Catalog Access Groups screen.

The matrix will look something like this:

![image](https://user-images.githubusercontent.com/8419000/167341793-763885a4-c622-4e18-82a8-027333a90eea.png)
In that screenshot, the catalogs are listed across the top of the matrix, the access groups on the side, and the checkboxes show which groups have access to which catalogs.  The colors are a quick indicator of where the checkboxes are.    I'm also thinking there can be a separate column after the last with a checkbox to delete a group.  Also a separate line at the bottom of the table to create a new group and mapping.

This matrix is a visualization of the catalog_access table:
```
+-----------------+---------------------+------+-----+---------+-------+
| Field           | Type                | Null | Key | Default | Extra |
+-----------------+---------------------+------+-----+---------+-------+
| access_group_id | int(11) unsigned    | NO   | MUL | NULL    |       |
| catalog_id      | int(11) unsigned    | NO   | MUL | NULL    |       |
| enabled         | tinyint(1) unsigned | NO   |     | 0       |       |
+-----------------+---------------------+------+-----+---------+-------+
```
Where access_group_id is the left column, catalog_id is the top row, and the enabled field indicates if a checkbox should be checked.

I also think a link on this new screen on the  "Show Catalogs" screen would be helpful.  Other future enhancements would be to have a multi-select dropdown in the Add Catalog screen to select to which access groups the catalog will be added on creation.

**Reply by troycarpenter** [reply]
Very crude redo of the matrix screen more as I described:
![image](https://user-images.githubusercontent.com/8419000/167344596-5ccb239d-a807-416d-84cb-dbbe15e6250c.png)

Of course it should fit visual in with the theme/layouts/etc and not look like a frankenstein.

**Reply by lachlan-00** [reply]
would be better to use a selectable list like these prefs which is in src/Module/Util/Ui.php
![image](https://user-images.githubusercontent.com/1305249/167344773-d28a5541-073c-4b98-b973-4d356ac54b58.png)

So you could do some kind of per-catalog list
![image](https://user-images.githubusercontent.com/1305249/167344884-9c4e880b-77df-40f2-b9b0-ccca4cd40991.png)

**Reply by lachlan-00** [reply]
but you've got it in groups now?

okay, this is looking heavy to resolve when i look at it.

There should be something to manage the groups and something to manage the users in the groups then a third page to assign the groups to catalogs.

**Reply by troycarpenter** [reply]
Is that for a user?  I'm not sure I understand where that's at.  Everything I've talked about so far is to get the definitions of the groups on one screen.  I do think the user screen should have a drop-down similar to the "User Access Level" drop down box, but is instead populated by the groups (but I'm leaning more to the "filter" terminology).

**Reply by troycarpenter** [reply]
GUI's are not really my thing, so I'll need more guidance. to make sure it all fits in.  All that matters is that the new tables get populated properly.  All the backend code is already using the new tables to do the filtering per user.  All that's left is to have a GUI to get the data into the tables and to be able to edit the data as people change their configurations.

**Reply by troycarpenter** [reply]
And now that I'm trying to add the selection for the filter on the Add user page, I realize I didn't create the lower level routines to fetch the data from those tables to return the values to front-end code.

**Reply by lachlan-00** [reply]
another question would be that visually the groups aren't needed in the UI

you could assign the users to a catalog and that access_group is updated based on the list of selected users

**Reply by lachlan-00** [reply]
maybe make a pull request and i can see how it all works together.

from what i'm reading i think i can work out a method for manipulating it easily.

**Reply by troycarpenter** [reply]
I can make a pull request, then you can see what's all happening in the background.  Be aware there is a one-time database action for upgrades to this feature that all catalogs will be added to the DEFAULT filter (or access) group.  Also, the feature is activated by the catalog_filter setting like the original single-user filter on a catalog.  This is done in case the catalog_filter option is ON so that everyone will still have access to catalogs upon upgrade and before the admin sets up his catalog filters.

**Reply by troycarpenter** [reply]
I'm working the leaf functions for setting and getting info out of the tables.  My first goal is to be able to set the appropriate filter via the add/edit user screens since those are simply fields in the user record.  Should be as simple as pulling the names out of the user_access_group table and building the dropdown selector with filter names to present with the values being the filter IDs.

The big task will be what is discussed above: the best way to get and present the user_access table to the user.  I have a idea for that which probably fits better into the GUI style.

**Comment by troycarpenter** [comment]
Question...As I start to write new functions, how do I test them to be sure they are working as expected?  It was easy to modify existing functions, but now I would like a simple way to call my routine as if it were are real application and see what the return values are.

I see the test folder, but I'm not entirely sure how to utilize it from the command line to test my leaf functions.

**Reply by troycarpenter** [reply]
I don't know the answer...I just hijacked some other functions that I knew would run and verified that way.

**Comment by troycarpenter** [comment]
After a lot longer than I wanted, I have the add user screen on the GUI able to insert the correct filter into the user profile:

![image](https://user-images.githubusercontent.com/8419000/167487865-c87cc885-6f93-4fa4-8c68-9dc74de06f6d.png)

Next I'll get the modify user working.

**Reply by troycarpenter** [reply]
Ugh, web gui work is tedious.  I've now got the edit user working.  I can freely change the catalog filter for a user, and the correct one is selected when going into the edit screen.

**Comment by troycarpenter** [comment]
Back to the Catalog Filter Creation screen.  To make it more like the other existing screens, I guess we could have a filter creation screen, and a "show filters" screen.  The create screen would allow the admin to put in a name for the filter, and a list of catalogs where the admin can check which catalogs are available to this filter.

The "show filters" screen could be similar to the Show Catalogs screen where a table of the existing filters are listed, with actions being to delete or edit the filters.  The edit filter would look similar to the add filter screen and allow the admin to change the name and which catalogs are available to that filter.

A future enhancement would be to allow which filters to which a catalog belongs to on it's creation or modification (either by checkbox or multi-select).

I would definitely want some help here, but I realize the time zone difference is pretty great.  I live in central Texas.

**Comment by troycarpenter** [comment]
I guess now is a good time to go over the architecture and what the lower level code expects.

This feature is a replacement for the catalog_filter functionality.  Instead of specifying a catalog that is private to a user,  the admin can create profiles that define which catalogs are available to a user.  The feature is turned on with the catalog_filter=true setting in the ampache.cfg.php file.

The access filter is stored in the catalog_access_group table.  It simply holds a unique ID for the filter, and a string name of the filter.  When a user is assigned a filter, the filter ID is stored in the catalog_access_group field in the user table.  A default group with ID=1 and name=DEFAULT is created.
```
MariaDB [ampache]> describe catalog_access_group;
+-------+------------------+------+-----+---------+----------------+
| Field | Type             | Null | Key | Default | Extra          |
+-------+------------------+------+-----+---------+----------------+
| id    | int(11) unsigned | NO   | PRI | NULL    | auto_increment |
| name  | varchar(128)     | YES  | UNI | NULL    |                |
+-------+------------------+------+-----+---------+----------------+
```
A second table, catalog_access, is also created which is a matrix of the filter ID and catalog ID, with a third field "enabled" which is 1 or 0 depending on if that filter/catalog combination is enabled for that profile.
```
MariaDB [ampache]> describe catalog_access;
+-----------------+---------------------+------+-----+---------+-------+
| Field           | Type                | Null | Key | Default | Extra |
+-----------------+---------------------+------+-----+---------+-------+
| access_group_id | int(11) unsigned    | NO   | MUL | NULL    |       |
| catalog_id      | int(11) unsigned    | NO   | MUL | NULL    |       |
| enabled         | tinyint(1) unsigned | NO   |     | 0       |       |
+-----------------+---------------------+------+-----+---------+-------+
```
Through database definitions, when a filter ID  is deleted, then all the corresponding rows are deleted from the catalog_access table.  Correspondingly, when a catalog ID is deleted, all rows with that catalog ID are deleted from the catalog_access table.  A fully filled catalog_access table would look like this:
With a catalog_access_group table with these entries:
```
MariaDB [ampache]> select * from catalog_access_group;
+----+---------+
| id | name    |
+----+---------+
|  1 | DEFAULT |
|  2 | GROUP 1 |
|  3 | GROUP 2 |
|  4 | Kids    |
+----+---------+
```
The access table could look like this:
```
MariaDB [ampache]> select * from catalog_access;
+-----------------+------------+---------+
| access_group_id | catalog_id | enabled |
+-----------------+------------+---------+
|               1 |          1 |       1 |
|               1 |          2 |       1 |
|               1 |          3 |       1 |
|               1 |          4 |       1 |
|               1 |          5 |       1 |
|               2 |          1 |       0 |
|               2 |          2 |       1 |
|               2 |          3 |       1 |
|               2 |          4 |       0 |
|               2 |          5 |       1 |
|               3 |          1 |       1 |
|               3 |          2 |       0 |
|               3 |          3 |       1 |
|               3 |          4 |       1 |
|               3 |          5 |       0 |
|               4 |          1 |       0 |
|               4 |          2 |       0 |
|               4 |          3 |       1 |
|               4 |          4 |       0 |
|               4 |          5 |       0 |
+-----------------+------------+---------+
```
Which is interpreted as follows:
Filter #1 (DEFAULT) has all catalogs enabled
Filter #2 (GROUP 1) enables access to catalogs 1,2, and 5 but not 1 or 4.
Filter #3 (GROUP 2) enables access to catalogs 1,3, and 4 but not 2 or 5.
Filter #4 (Kids) only allows access to catalog 3.

The requirements are as follows:
1.  When adding a new filter, an entry must be created for each access_group_id, catalog_id combination with zeros in all the newly created rows "enabled" field.
2. When adding a catalog, a new entry must be created for each access_group_id, catalog_id combination with zeros in all the newly created rows "enabled" field.

As stated earlier, the delete relationships are handled by the database.  When a filter is deleted which has been assigned to a user or group of users, then those users will have their referenced filter be reset to DEFAULT.

Please let me know if you have any questions.  If you don't get around to starting the new GUI code, I'll take a crack at it soon.

**Reply by troycarpenter** [reply]
I think I may have figured out the GUI structure.  I was able to create a new "Add Filter" screen and now need to create the logic and see if it will actually put the values in the database.

**Reply by troycarpenter** [reply]
Here's the add screen for a filter:

![image](https://user-images.githubusercontent.com/8419000/167675390-a4d7f26c-2041-41e4-ab7d-b7e99ecab7dd.png)
It detects if there is no filter name, and if the filter name already exists.
I need to add actually creating the filter based on the requirements I listed above.

**Reply by troycarpenter** [reply]
I now have it creating all the database entries correctly.  Going to move to the browsing and editing screens now.

**Reply by troycarpenter** [reply]
This is the browse screen, which really needs work because I couldn't figure out all the fancy table adornments that are configured somewhere for the other table types:
![image](https://user-images.githubusercontent.com/8419000/167723400-43182a01-cc39-4dca-8eed-4661a0e5abe7.png)
This is more of a mockup, although on a live system.  I populated static dummy info for the filters so I could fill out the table.  The number stats are static right now, and I want to change the action section to be the icons (gear for edit, X for delete) instead of dropdown selects.

I'm trying to get as much done before my other duties (job, family) start asking why I'm not getting other tasks done.

**Reply by troycarpenter** [reply]
I'm also creating a bit of a mess in src/Repository/Model/Catalog.php,,,I've been adding my utility functions there.

**Reply by troycarpenter** [reply]
Got it pulling correct data now:
![image](https://user-images.githubusercontent.com/8419000/167770764-3b8c19a6-0f9d-4329-a8a8-4b1020050634.png)
But still missing all the table decorations that go with most tables.

**Comment by raflz** [comment]
It would also be convenient to set permissions directly from the album screen, as is done now for individual songs (but there is only a general ban on showing, excluding groups).
![image](https://user-images.githubusercontent.com/104421534/167772856-7ea464bd-b9f6-44fb-a417-42fce8295ab1.png)

**Reply by troycarpenter** [reply]
I'm not sure this feature is addressing permissions.  It's more of a big 4" paint brush instead of a fine pencil.  With this feature, if a song is in a catalog that is not allowed, then the song will not even show up in the interface anywhere.

**Reply by raflz** [reply]
I suggested considering creating a filter call screen directly from the album (similar to a song).

**Reply by troycarpenter** [reply]
This may not be the same filters you are thinking.  However, there ~~is~~ will be a link to the filters on the left menu, one for "add catalog filter" and another for the browse screen above.

**Reply by troycarpenter** [reply]
My goal right now is to get this feature submitted and functional, then improvements can be made.  If your suggestion can fit into this feature then I'm sure it can eventually be put there.  But really it sounds like you are suggesting an enhancement to song and album management. which may be a different feature/enhancement request.

**Reply by troycarpenter** [reply]
I have looked at that screen you referenced.  It would appear that icon toggles the disabled flag for a track.  My guess is that it will disable the track for everyone.

The feature described here operates on the catalog level.  Song and album availably are evaluated based on which catalog the song and/or album belongs, and then the catalogs to which a user has access.

For instance, I have a folder in my collection called "Kids" and I create a catalog called Kids pointing to that directory.  I then create a catalog filter also called "Kids" (for simplicity, but it can be named anything), which only allows access to the Kids catalog.  Then, for the user accounts that belong to my children, I set the assigned filter to be Kids.  From that point on, they only have access to songs/albums/artists that are in the Kids catalog.  If an artist also has songs/albums in another catalog, the artist is shown with the songs/albums that are in the Kids catalog, but any songs/albums that are in another catalog get filtered out and are never seen.

This actually brings Ampache up to par with other *sonic music servers in terms of catalog access for users, and is the main reason I haven't switched my users from Airsonic Advanced to Ampache yet.  Once this gets merged, I will swap out AA and change all my DNS pointers to Ampache and hope it's rather transparent for the mobile users (using Dsub on Android).

**Comment by troycarpenter** [comment]
As of tonight, I have the Add filter GUI, the incomplete browse table of filters, and a delete filter screen which is access from the browse table..

TODO:
~~Create add filter screen~~
~~Create browse table with Edit/Delete links~~ Note: this is functional but not visually there yet.
~~Create delete filter code~~
Create edit filter code
Add routines to catalog creation to add a new catalog to the catalog_access table.
Clean up code and submit final pull request.

**Reply by troycarpenter** [reply]
One thing I'd like to figure out is this.  When I do error testing on my form, it redraws the form with no pre-filled data.  For instance, on an edit, all the fields are populated with the current data for the user.  If I do something that generates an error like try to duplicate a filter name, then I get a message the name is already taken, but then that field blanks and all the other pre-filled data disappears (namely the checkboxes that indicate which catalogs are active).  Is there any way to get the form to refresh after the error with the pre-filled data?  I have not tried it on other forms in the system to see if the behavior is the same.

**Reply by troycarpenter** [reply]
The modify code is now done.  It is now possible to change the catalogs in the filter, as well as rename the filter.

**Comment by troycarpenter** [comment]
I have now fully committed the code to the code repository and it is available in pull request #3260.

All code is functional, but there are still some cosmetic things that need to be addressed, and the back-end code needs to be reorganized.

**Comment by mitchray** [comment]
Great work on this @troycarpenter!

**Reply by troycarpenter** [reply]
> Great work on this @troycarpenter!

Thank you!

The absence of this feature was keeping me from moving my users from a *sonic based server to Ampache.  And since all the other *sonic type servers have pretty much gone stale, I really wanted to get my users onto Ampache.  So in that spirit, I went ahead and implemented this feature.

By the way, I've been looking at Ample and will be suggesting it to my users for music.  We are heavy podcast listeners also, so there would be back-and-forth on the web interface.  For mobile, everyone pretty much uses Dsub (especially since it's now available for free).
